File: ruby/lib/runapi/infinitetalk/resources/audio_to_video.rb
# frozen_string_literal: true

module RunApi
  module Infinitetalk
    module Resources
      # Generates lip-synced talking-head videos from a portrait image and an audio track.
      # The output video shows the person speaking or singing in sync with the audio.
      class AudioToVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/infinitetalk/audio_to_video"
        RESPONSE_CLASS = Types::AudioToVideoResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedAudioToVideoResponse
        PROMPT_MAX_LENGTH = 5000
        SEED_RANGE = (10_000..1_000_000)

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["audio-to-video"], params)

          prompt = param(params, :prompt)
          raise Core::ValidationError, "prompt is required" unless prompt.is_a?(String) && !prompt.empty?
          if prompt.length > PROMPT_MAX_LENGTH
            raise Core::ValidationError, "prompt must be at most #{PROMPT_MAX_LENGTH} characters"
          end

          seed = param(params, :seed)
          return if seed.nil?

          seed = Integer(seed, exception: false)
          return if seed && SEED_RANGE.cover?(seed)

          raise Core::ValidationError, "seed must be an integer between #{SEED_RANGE.min} and #{SEED_RANGE.max}"
        end
      end
    end
  end
end

File: python/src/runapi/infinitetalk/resources/audio_to_video.py
"""InfiniteTalk audio-to-video resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource, ValidationError

from ..contract_gen import CONTRACT
from ..types import (
    AudioToVideoResponse,
    CompletedAudioToVideoResponse,
)

PROMPT_MAX_LENGTH = 5000
SEED_RANGE = range(10_000, 1_000_001)


class AudioToVideo(Resource):
    """Generate a talking-head video from a source image and audio track."""

    ENDPOINT = "/api/v1/infinitetalk/audio_to_video"

    RESPONSE_CLASS = AudioToVideoResponse
    COMPLETED_RESPONSE_CLASS = CompletedAudioToVideoResponse

    def run(self, **params: Any) -> Any:
        """Create a task and poll until it completes."""
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create an audio-to-video task and return immediately with an ``id``."""
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of an audio-to-video task."""
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["audio-to-video"], params)

        prompt = params.get("prompt")
        if not (isinstance(prompt, str) and prompt != ""):
            raise ValidationError("prompt is required")
        if len(prompt) > PROMPT_MAX_LENGTH:
            raise ValidationError(f"prompt must be at most {PROMPT_MAX_LENGTH} characters")

        seed = params.get("seed")
        if seed is None:
            return

        parsed = self._parse_integer(seed)
        if parsed is not None and parsed in SEED_RANGE:
            return

        raise ValidationError(
            f"seed must be an integer between {SEED_RANGE.start} and {SEED_RANGE.stop - 1}"
        )

    @staticmethod
    def _parse_integer(value: Any) -> Any:
        """Mirror Ruby ``Integer(value, exception: false)``.

        Accepts ints and integer-looking strings; rejects bools and anything
        non-coercible by returning ``None``.
        """
        if isinstance(value, bool):
            return None
        if isinstance(value, int):
            return value
        if isinstance(value, str):
            try:
                return int(value.strip(), 10)
            except ValueError:
                return None
        return None
